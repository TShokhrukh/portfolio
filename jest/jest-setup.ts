import '@testing-library/jest-dom'
import { TextEncoder, TextDecoder } from 'util'
import { IntersectionObserverMock } from './mock/IntersectionObserverMock '

// @ts-ignore
global.TextEncoder = TextEncoder
// @ts-ignore
global.TextDecoder = TextDecoder
// @ts-ignore
global.IntersectionObserver = IntersectionObserverMock
