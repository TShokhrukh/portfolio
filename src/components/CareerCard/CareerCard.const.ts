import { cn } from '@bem-react/classname'

export const cnCareerCard = cn('CareerCard')

export const dateFormat: string = 'yyyy/LLLL'
