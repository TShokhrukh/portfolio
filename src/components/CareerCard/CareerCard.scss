.CareerCard_active {
  --card-border-color: transparent;
  background-color: rgba($purple, 0.1);
}
