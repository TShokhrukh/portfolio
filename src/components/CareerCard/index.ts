export * from './build'
export * from './__types__'
export { CareerCard as CareerCardBase } from './CareerCard'
