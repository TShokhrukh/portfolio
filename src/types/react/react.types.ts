import type { ElementType } from 'react'

export interface IComponentProp<P = any> {
  component?: ElementType<P>,
}
