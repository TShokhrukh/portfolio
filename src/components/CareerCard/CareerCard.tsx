import React, { FC } from 'react'
import { useTranslator } from '@eo-locale/react'
import { Card } from '~/components/Card/build'
import { cnCareerCard } from './CareerCard.const'
import { Context } from './lib/context'
import type { ICareerCardProps } from './__types__'
import { CareerCardTitle } from './Title'
import { CareerCardDesc } from './Desc'
import { CareerCardWorking } from './Working'
import './CareerCard.scss'

export const CareerCard: FC<ICareerCardProps> = ({ career, className, ...props }) => {
  const t = useTranslator()
  return (
    <Context.Provider value={{ career }}>
      <Card {...props} className={cnCareerCard({ active: career.active }, [className])}>
        <CareerCardTitle>{career.title}</CareerCardTitle>
        <CareerCardDesc>{career.desc[t.language]}</CareerCardDesc>
        <CareerCardWorking />
      </Card>
    </Context.Provider>
  )
}
