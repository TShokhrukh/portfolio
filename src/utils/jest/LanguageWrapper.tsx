import React, { ReactNode } from 'react'
import { TranslationsProvider } from '@eo-locale/react'

import type { Locale } from '@eo-locale/core'

interface IWrapperProps {
  children: ReactNode,
  language?: string,
  locales?: Locale[]
}

const defLocales: Locale[] = [
  {
    language: 'en',
    messages: {}
  }
]

export const LanguageWrapper: React.FC<IWrapperProps> = ({
  children,
  language = 'en',
  locales = defLocales
}) => {
  return (
    <TranslationsProvider language={language} locales={locales} onError={() => null}>
      {children}
    </TranslationsProvider>
  )
}
