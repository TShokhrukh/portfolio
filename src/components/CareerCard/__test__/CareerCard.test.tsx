/* global test, expect, describe */
import React from 'react'
import intervalToDuration from 'date-fns/intervalToDuration'
import parseISO from 'date-fns/parseISO'
import { render, screen } from '@testing-library/react'
import { CareerCard } from '../CareerCard'
import { LanguageWrapper } from '~/utils/jest/LanguageWrapper'
import type { ICareer } from '../__types__'

const locales = [
  {
    language: 'en',
    messages: {
      careerCard: {
        durationYears: '{years} Y',
        durationMonths: '{months} M'
      }
    }
  }
]

describe('CareerCard', () => {
  test('must return the component template', () => {
    const career: ICareer = {
      id: '1',
      active: false,
      title: 'Title',
      desc: {
        en: 'desc'
      },
      startDate: '2020-04-09',
      endDate: '2023-06-04'
    }
    render(
      <LanguageWrapper locales={locales}>
        <CareerCard career={career} data-testid='career_card' />
      </LanguageWrapper>
    )

    expect(screen.getByTestId('career_card')).toMatchSnapshot()
  })

  test('must have \'CareerCard_active\' if active=true for the career.', () => {
    const career: ICareer = {
      id: '1',
      active: true,
      title: 'Title',
      desc: {
        en: 'desc'
      },
      startDate: '2020-04-09',
      endDate: '2023-06-04'
    }
    render(
      <LanguageWrapper locales={locales}>
        <CareerCard career={career} data-testid='career_card' />
      </LanguageWrapper>
    )

    expect(screen.getByTestId('career_card')).toHaveAttribute('class', 'Card CareerCard CareerCard_active')
  })

  test('must throw error when the end date is greater than the start date', () => {
    const career: ICareer = {
      id: '1',
      active: true,
      title: 'Title',
      desc: {
        en: 'desc'
      },
      startDate: '2020-04-09',
      endDate: '2002-06-04'
    }

    expect(() => {
      render(
        <LanguageWrapper locales={locales}>
          <CareerCard career={career} data-testid='career_card' />
        </LanguageWrapper>
      )
    }).toThrowError('endDate cannot be earlier than startDate')
  })

  test('by default, it must take today\'s date for \'endDate\' if it is not provided', () => {
    const startDate = '2020-04-09'
    const career: ICareer = {
      id: '1',
      active: false,
      title: 'Title',
      desc: {
        en: 'desc'
      },
      startDate,
      endDate: ''
    }
    const duration = intervalToDuration({
      start: parseISO(startDate),
      end: new Date()
    })
    const content = `${duration.years} Y` + (duration.months ? `${duration.months} M` : '')

    render(
      <LanguageWrapper locales={locales}>
        <CareerCard career={career} data-testid='career_card' />
      </LanguageWrapper>
    )

    expect(screen.getByText(content)).toBeInTheDocument()
  })
})
