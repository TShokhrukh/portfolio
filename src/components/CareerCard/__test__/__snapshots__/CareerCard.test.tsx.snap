// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CareerCard must return the component template 1`] = `
<div
  class="Card CareerCard"
  data-testid="career_card"
>
  <h5
    class="Typography Typography_color_white Typography_variant_h5 CareerCard-Title"
  >
    Title
  </h5>
  <p
    class="Typography CareerCard-Desc"
  >
    desc
  </p>
  <div
    class="CareerCard-Working"
  >
    <p
      class="Typography Typography_color_purple CareerCard-Duration"
    >
      3
       Y
      1
       M
    </p>
    <p
      class="Typography Typography_color_purple CareerCard-Date"
    >
      2020/April
       - 
      2023/June
    </p>
  </div>
</div>
`;
